File: examples/bankaccount-springboot-java/src/main/java/io/toolisticon/axon/avro/examples/bankaccount/rest/BankAccountController.java
package io.toolisticon.axon.avro.examples.bankaccount.rest;

import io.toolisticon.axon.avro.examples.bankaccount.command.CreateBankAccountCommand;
import io.toolisticon.axon.avro.examples.bankaccount.command.DepositMoneyCommand;
import io.toolisticon.axon.avro.examples.bankaccount.command.WithdrawMoneyCommand;
import io.toolisticon.axon.avro.examples.bankaccount.projection.CurrentBalanceProjection;
import io.toolisticon.axon.avro.examples.bankaccount.projection.CurrentBalanceProjection.BankAccountInfoDto;
import io.toolisticon.axon.avro.examples.bankaccount.projection.CurrentBalanceProjection.FindAll;
import io.toolisticon.axon.avro.examples.bankaccount.projection.CurrentBalanceProjection.FindById;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.queryhandling.QueryGateway;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

import static org.axonframework.messaging.responsetypes.ResponseTypes.multipleInstancesOf;
import static org.axonframework.messaging.responsetypes.ResponseTypes.optionalInstanceOf;

@RestController
@RequestMapping("/api/bankaccounts")
public class BankAccountController {

  private final CommandGateway commandGateway;
  private final QueryGateway queryGateway;

  public BankAccountController(CommandGateway commandGateway, QueryGateway queryGateway) {
    this.commandGateway = commandGateway;
    this.queryGateway = queryGateway;
  }

  @PostMapping("/{bankAccountId}")
  public ResponseEntity<Void> create(
    @PathVariable("bankAccountId") String bankAccountId,
    @RequestParam(value = "initialBalance", required = false) Integer initialBalance
  ) {
    commandGateway.sendAndWait(
      new CreateBankAccountCommand(
        bankAccountId,
        Optional.ofNullable(initialBalance).orElse(0))
    );

    return ResponseEntity.ok().build();
  }

  @PutMapping("/{bankAccountId}")
  public ResponseEntity<Void> update(
    @PathVariable("bankAccountId") String bankAccountId,
    @RequestParam(value = "amount") int amount
  ) {
    if (amount == 0) {
      throw new IllegalArgumentException("amount must be non zero");
    } else if (amount > 0) {
      commandGateway.sendAndWait(
        new DepositMoneyCommand(
          bankAccountId,
          amount
        )
      );
    } else {
      commandGateway.sendAndWait(
        new WithdrawMoneyCommand(
          bankAccountId,
          -amount
        )
      );
    }

    return ResponseEntity.ok().build();
  }

  @GetMapping("/{bankAccountId}")
  public ResponseEntity<BankAccountInfoDto> findById(
    @PathVariable("bankAccountId") String bankAccountId
  ) {
    return ResponseEntity.of(
      queryGateway.query(
        new FindById(bankAccountId),
        optionalInstanceOf(BankAccountInfoDto.class)
      ).join()
    );
  }

  @GetMapping("/")
  public ResponseEntity<List<BankAccountInfoDto>> findAll() {
    return ResponseEntity.ok(
      queryGateway.query(
        new FindAll(),
        multipleInstancesOf(BankAccountInfoDto.class)
      ).join()
    );
  }

}

File: examples/bankaccount-springboot-kotlin/src/main/kotlin/BankAccountController.kt
package io.holixon.axon.avro.examples.bankaccount

import bankaccount.command.CreateBankAccount
import bankaccount.command.DepositMoney
import bankaccount.command.WithdrawMoney
import bankaccount.projection.CurrentBalanceProjection
import org.axonframework.commandhandling.gateway.CommandGateway
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/bankaccounts")
class BankAccountController(
  private val commandGateway: CommandGateway,
  private val currentBalanceQueries: CurrentBalanceProjection.CurrentBalanceQueries
) {

  @PostMapping("/{bankAccountId}")
  fun create(
    @PathVariable("bankAccountId") bankAccountId: String,
    @RequestParam(value = "initialBalance", required = false) initialBalance: Int?
  ): ResponseEntity<Void> {
    commandGateway.sendAndWait<CreateBankAccount>(CreateBankAccount(
      bankAccountId, initialBalance ?: 0
    ))
    return ResponseEntity.ok().build()
  }

  @PutMapping("/{bankAccountId}")
  fun update(
    @PathVariable("bankAccountId") bankAccountId: String?,
    @RequestParam(value = "amount") amount: Int
  ): ResponseEntity<Void?>? {
    require(amount != 0) { "amount must be non zero" }
    if (amount > 0) {
      commandGateway.sendAndWait<DepositMoney>(
        DepositMoney(
          bankAccountId,
          amount
        )
      )
    } else {
      commandGateway.sendAndWait<WithdrawMoney>(WithdrawMoney(bankAccountId, -amount))
    }
    return ResponseEntity.ok().build()
  }

  @GetMapping("/{bankAccountId}")
  fun findById(
    @PathVariable("bankAccountId") bankAccountId: String
  ): ResponseEntity<CurrentBalanceProjection.CurrentBalance> {
    return ResponseEntity.of(
      currentBalanceQueries.findByAccountId(bankAccountId).join()
    )
  }

  @GetMapping("/")
  fun findAll(): ResponseEntity<List<CurrentBalanceProjection.CurrentBalance>> {
    return ResponseEntity.ok(
      currentBalanceQueries.findAll().join()
    )
  }
}
